internal/blocktx/store/postgresql/upsert_block_transactions.go

```diff
@@ -2,109 +2,100 @@ package postgresql
 
 import (
 	"context"
-	"encoding/hex"
 	"fmt"
 
-	"github.com/bitcoin-sv/arc/internal/blocktx/blocktx_api"
 	"github.com/bitcoin-sv/arc/internal/blocktx/store"
 	"github.com/lib/pq"
 	"go.opentelemetry.io/otel/trace"
 )
 
 // UpsertBlockTransactions upserts the transaction hashes for a given block hash and returns updated registered transactions hashes.
-func (p *PostgreSQL) UpsertBlockTransactions(ctx context.Context, blockId uint64, transactions []*blocktx_api.TransactionAndSource, merklePaths []string) ([]store.UpsertBlockTransactionsResult, error) {
+func (p *PostgreSQL) UpsertBlockTransactions(ctx context.Context, blockId uint64, txsWithMerklePaths []store.TxWithMerklePath) ([]store.TxWithMerklePath, error) {
 	if tracer != nil {
 		var span trace.Span
 		ctx, span = tracer.Start(ctx, "UpdateBlockTransactions")
 		defer span.End()
 	}
 
-	if len(transactions) != len(merklePaths) {
-		return nil, fmt.Errorf("transactions (len=%d) and Merkle paths (len=%d) have not the same lengths", len(transactions), len(merklePaths))
+	blockIDs := make([]uint64, len(txsWithMerklePaths))
+	txHashesBytes := make([][]byte, len(txsWithMerklePaths))
+	merklePaths := make([]string, len(txsWithMerklePaths))
+	for i, tx := range txsWithMerklePaths {
+		blockIDs[i] = blockId
+		txHashesBytes[i] = tx.Hash
+		merklePaths[i] = tx.MerklePath
 	}
 
-	txHashes := make([][]byte, len(transactions))
-	txHashesMap := map[string]int{}
-	for pos, tx := range transactions {
-		txHashes[pos] = tx.Hash
-		txHashesMap[hex.EncodeToString(tx.Hash)] = pos
-	}
+	qUpsertTransactions := `
```

Comment:
I would suggest reimplementing the query to something like this:
```sql
WITH inserted_transactions AS (
    INSERT INTO transactions (hash)
    SELECT UNNEST($2::BYTEA[])
    ON CONFLICT (hash) DO NOTHING
    RETURNING id, hash
)

INSERT INTO block_transactions_map (blockid, txid, merkle_path)
SELECT 
    $1::BIGINT,
    UNNEST(it.id, $3::TEXT[])
FROM inserted_transactions it
JOIN LATERAL UNNEST($2::BYTEA[]) AS t(hash) ON it.hash = t.hash;
```
and in go:
```go
dbTx.ExecContext(ctx, query, blockId, txHashesBytes, merklePaths)
```
Thanks to this:
* we do it all in one round trip
* we don't need to give up the SERIAL PK on the transactions table

_Disclaimer: I did not test my SQL statement, so there is a chance I’m totally wrong._

Reply:
This is a great idea, with some small improvements, I added this query to the UpsertTransactions method. Thanks a lot!